### PR TITLE
Improving CI/CD for Github actions

### LINKED ISSUE #202: Implement Setuptools-scm
[Setuptool-scm](https://pypi.org/project/setuptools-scm/) incorporates a git sha into every package number making versioning even more reliable. 

To implement:
1. add as build dependency
2. Add necessary metadata
3. Add Manifest file to exclude excess files.

### LINKED ISSUE #326: Improve CI/CD
Features to add:

1. [x] Coverage reports. Possibly coverall
2. [x] Add badges
3.  [x] Implement auto releases
4.  [x] Test for undocumented api packages now that they are not automatically generated. 
5.  [x] ~implement linting~ #337
6.  [x] add PR template (not CD per se)

Probably implement along with #202.

### LINKED ISSUE #328: Sentence fragment in README
> Can write out the MCNP problem verbatim, and try to match

I believe this should be "try to match user formatting."

### LINKED ISSUE #329: Go all in on pyproject
**Is your feature request related to a problem? Please describe.**
Currently project meta data are stored in four files: `pyproject.toml`, `pytest.ini`, and `requirements/*.txt`. 

**Describe the solution you'd like**
Consolidate all meta data into `pyproject.toml`.  The `pytest.ini` consolidation should be very simple. Move all requirements information to `pyproject.toml` is not as simple. This would be eliminating concrete dependencies and switching to purely abstract dependencies. This is not appropriate for all projects, but for how simple the dependencies are for MontePy, this should be possible.

**Additional context**
Some discussions about the possibility of this transition:

* [github discussion](https://github.com/pypa/packaging.python.org/issues/685)
* [Official guidance on dependencies vs. requirements](https://packaging.python.org/en/latest/discussions/install-requires-vs-requirements/)
* [Official discussion of concrete vs. abstract](https://caremad.io/posts/2013/07/setup-vs-requirement/)

This would require updating CI/CD because `pip install -r` will need to be removed. 

TODO list:

* [x] move over pytest
* [x] Move over dependencies

### LINKED ISSUE #338: Add DOI information to user docs
MontePy now has it's own DOI: https://doi.org/10.11578/dc.20240115.1 . We should probably add documentation as well for a "publications" section.

### PR DESCRIPTION
**Pull request by MicahGale**
Improvements: 

* Moved completely to pyproject.toml
   * deleted pytest.ini
   * delete `requirements/*`
* Added coveralls support
* Added a test report summary
* Went completely to git based `setuptools-scm` for controlling version
  * version system is robust and should never lead to an error
  * versions are determined by a git tag
  * Git tag is automatically created with an action that uses semantic versioning to decide if it's a patch, minor, or major release.
 * Added badges
 * Improved deploy process.
    * implemented actual build-deploy process
    * build will:
        1. bump to a new version, and create a git tag
        2. build
        3. sign the packages with sig store
        4. Create a draft release on GitHub
        5. Upload artifacts for later deployment to PyPi
 * Created a Pull Request Template

TODO:

* [x] enable coveralls for this repo
* [x] Add coveralls as a collaborator
* [x] document how semantic versioning works
* [x] Finish implementing #326. 
* [x] Fix #328.
* [x] Fix #338. 

resolves #326, resolves #202, resolves #329, resolves #328, resolves #338.

### CODE REVIEW

**Comment by MicahGale** [comment]
So the one failed test is due to `test_reporter` failing. This is a permissions issue. You can see this succeeding here: https://github.com/MicahGale/MontePy/actions/runs/7660917606. I suspect once this branch is in this repo, it will work fine.

Likewise Coveralls needs to run on deploy before it really works. You can see this here: https://github.com/MicahGale/MontePy/pull/2

For seeing the deploy process work see: https://github.com/MicahGale/MontePy/actions/runs/7605663330 this does fail because I was not going to allow it to deploy to PyPI

**Comment by MicahGale** [comment]
* [ ] bug issue template doesn't add bugs label.

**Comment by MicahGale** [comment]
Could you actually get to this one first?

File: .github/workflows/deploy.yml [hunk 1]
```diff
@@ -4,6 +4,9 @@ on:
   push:
     branches: [main]
 
+env:
+  RELEASE_LEVEL: patch
+
```
Comment (MicahGale) [inline .github/workflows/deploy.yml:9]: This is no longer needed. Should I remove it?

File: doc/source/api/montepy.data_inputs.rst
Comment (tjlaboss) [inline doc/source/api/montepy.data_inputs.rst:1]: Does `doc/source/api/` need to be included in the repository? I always build it with `sphinx-apidoc` as part of CI/CD.

File: doc/source/publications.rst
Comment (tjlaboss) [inline doc/source/publications.rst:1]: This is nice

File: montepy/__main__.py [hunk 2]
```diff
@@ -59,8 +65,10 @@ def main():  # pragma: no cover
     The main function
     """
     args = define_args(sys.argv[1:])
```
Comment (tjlaboss) [inline montepy/__main__.py:67]: Touching `sys.argv` should never be necessary when using `argparse`

File: montepy/__main__.py [hunk 3]
```diff
@@ -59,8 +65,10 @@ def main():  # pragma: no cover
...
-    if "check" in args:
+    if "check" in args and args.check:
```
Comment (tjlaboss) [inline montepy/__main__.py:68]:
```python
if args.check:
```
should be sufficient

File: montepy/data_inputs/fill.py [hunk 4]
```diff
@@ -525,9 +525,11 @@ def _value_node_generator():
                         payload.append(self.universes[i][j][k].number)
         else:
             payload = [
-                self.universe.number
-                if self.universe is not None
-                else self.old_universe_number
+                (
```
Comment (tjlaboss) [inline montepy/data_inputs/fill.py:528]: Are these parens necessary?

File: .github/workflows/main.yml
Comment (tjlaboss) [inline .github/workflows/main.yml:1]: I will have to learn more about GitHub Actions to review this appropriately.

File: .github/workflows/deploy.yml
Comment (tjlaboss) [inline .github/workflows/deploy.yml:1]: I will have to learn more about GitHub Actions

**Review by tjlaboss — CHANGES_REQUESTED** [review]

File: montepy/__main__.py (hunk 2 above)
Comment (MicahGale) [reply montepy/__main__.py:67]: Oh really? I'll look into the preferred way.

File: montepy/__main__.py (hunk 3 above)
Comment (MicahGale) [reply montepy/__main__.py:68]: Agreed.

File: .github/workflows/main.yml
Comment (MicahGale) [reply .github/workflows/main.yml:1]: Yeah... I am very cargo cult. I'd say just look at the past actions that ran in my repo.

File: montepy/data_inputs/fill.py (hunk 4 above)
Comment (MicahGale) [reply montepy/data_inputs/fill.py:528]: That's added by `black` so I'm not messing with it. I think it does make it more legible though expands it.

**Comment by MicahGale** [comment]
Post Merge Todo:
* [ ] enforce passing checks for all PRs to develop and main

**Pull request merged**